net/multi_listen.go

```diff
@@ -0,0 +1,182 @@
+limitations under the License.
+*/
+
+package net
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+)
+
+// connErrPair pairs conn and error which is returned by accept on sub-listeners.
+type connErrPair struct {
+	conn net.Conn
+	err  error
+}
+
+// multiListener implements net.Listener
+type multiListener struct {
+	listeners []net.Listener
+
+	wg     sync.WaitGroup
+	mu     sync.Mutex
+	closed bool
+
+	// connErrQueue holds the connections accepted by sub-listeners
+	connErrQueue []connErrPair
+
+	// acceptReadyCh is used as a semaphore to wake up the waiting
+	// multiListener.Accept() when new connections are available
+	acceptReadyCh chan any
+}
+
+// compile time check to ensure *multiListener implements net.Listener
+var _ net.Listener = &multiListener{}
+
+// MultiListen returns net.Listener which can listen on and accept connections for
+// the given network on multiple addresses. Internally it uses stdlib to create
+// sub-listener and multiplexes connection requests using go-routines.
+// The network must be "tcp", "tcp4" or "tcp6".
+// It follows the semantics of net.Listen that primarily means:
+//  1. If the host is an unspecified/zero IP address with "tcp" network, MultiListen
+//     listens on all available unicast and anycast IP addresses of the local system.
+//  2. Use "tcp4" or "tcp6" to exclusively listen on IPv4 or IPv6 family, respectively.
+//  3. The host can accept names (e.g, localhost) and it will create a listener for at
+//     most one of the host's IP.
+func MultiListen(ctx context.Context, network string, addrs []string) (net.Listener, error) {
+	return multiListen(
+		ctx,
+		network,
+		addrs,
+		func(ctx context.Context, network, address string) (net.Listener, error) {
+			var lc net.ListenConfig
+			return lc.Listen(ctx, network, address)
+		})
+}
+
+// multiListen implements MultiListen by consuming stdlib functions as dependency allowing
+// mocking for unit-testing.
+func multiListen(
+	ctx context.Context,
+	network string,
+	addrs []string,
+	listenFunc func(ctx context.Context, network, address string) (net.Listener, error),
+) (net.Listener, error) {
+	if !(network == "tcp" || network == "tcp4" || network == "tcp6") {
+		return nil, fmt.Errorf("network '%s' not supported", network)
+	}
+	if len(addrs) == 0 {
+		return nil, fmt.Errorf("no address provided to listen on")
+	}
+
+	ml := &multiListener{
+		acceptReadyCh: make(chan any),
+	}
+
+	for _, addr := range addrs {
+		l, err := listenFunc(ctx, network, addr)
+		if err != nil {
+			// close all the sub-listeners and exit
+			_ = ml.Close()
+			return nil, err
+		}
+		ml.listeners = append(ml.listeners, l)
+	}
+
+	for _, l := range ml.listeners {
+		ml.wg.Add(1)
+		go func(l net.Listener) {
+			defer ml.wg.Done()
+			for {
+				conn, err := l.Accept()
+				ml.mu.Lock()
+				if ml.closed {
+					ml.mu.Unlock()
+					return
+				}
+				// enqueue the accepted connection
+				ml.connErrQueue = append(ml.connErrQueue, connErrPair{conn: conn, err: err})
```

Comment:
> This means the queue is unbounded? Is that really what we want?

No, probably not.

But your version above has the race condition: it's possible for `ml.Close()` to be called in between the `if closed { return }` and the `ch <- connErrPair{...` in the listening thread, resulting in a panic.

Reply:
> But as far as I can tell there is no non-blocking `Accept()` in Go (unless maybe we fiddle with the socket FD which seems dangerous).

(and not Windows-compatible)

Reply:
Indeed, that was bad.  This may be better.  Gets rid of the mutex entirely, at the cost of one more channel, but follows the Go ideal of "senders close channels".

```diff
diff --git a/net/multi_listen.go b/net/multi_listen.go
index d508805..6a94f46 100644
--- a/net/multi_listen.go
+++ b/net/multi_listen.go
@@ -29,20 +29,28 @@ type connErrPair struct {
 	err  error
 }
 
+func (p connErrPair) String() string {
+	if p.conn != nil {
+		return p.conn.RemoteAddr().String()
+	}
+	if p.err != nil {
+		return p.err.Error()
+	}
+	return "neither conn nor err were valid"
+}
+
 // multiListener implements net.Listener
 type multiListener struct {
 	listeners []net.Listener
 
-	wg     sync.WaitGroup
-	mu     sync.Mutex
-	closed bool
+	wg sync.WaitGroup
 
-	// connErrQueue holds the connections accepted by sub-listeners
-	connErrQueue []connErrPair
-
-	// acceptReadyCh is used as a semaphore to wake up the waiting
-	// multiListener.Accept() when new connections are available
-	acceptReadyCh chan any
+	// connChs passes accepted connections, from child listeners to parent.
+	connChs []chan connErrPair
+	// readyCh indicates that a child listener's connCh has data.
+	readyCh chan struct{}
+	// stopCh communicates from parent to child listeners.
+	stopCh chan struct{}
 }
 
 // compile time check to ensure *multiListener implements net.Listener
@@ -58,13 +66,13 @@ var _ net.Listener = &multiListener{}
 //  2. Use "tcp4" or "tcp6" to exclusively listen on IPv4 or IPv6 family, respectively.
 //  3. The host can accept names (e.g, localhost) and it will create a listener for at
 //     most one of the host's IP.
-func MultiListen(ctx context.Context, network string, addrs []string) (net.Listener, error) {
+func MultiListen(ctx context.Context, network string, addrs ...string) (net.Listener, error) {
+	var lc net.ListenConfig
 	return multiListen(
 		ctx,
 		network,
 		addrs,
 		func(ctx context.Context, network, address string) (net.Listener, error) {
-			var lc net.ListenConfig
 			return lc.Listen(ctx, network, address)
 		})
 }
@@ -78,14 +86,16 @@ func multiListen(
 	listenFunc func(ctx context.Context, network, address string) (net.Listener, error),
 ) (net.Listener, error) {
 	if !(network == "tcp" || network == "tcp4" || network == "tcp6") {
-		return nil, fmt.Errorf("network '%s' not supported", network)
+		return nil, fmt.Errorf("network %q not supported", network)
 	}
 	if len(addrs) == 0 {
 		return nil, fmt.Errorf("no address provided to listen on")
 	}
 
 	ml := &multiListener{
-		acceptReadyCh: make(chan any),
+		connChs: make([]chan connErrPair, 0, len(addrs)),
+		readyCh: make(chan struct{}),
+		stopCh:  make(chan struct{}),
 	}
 
 	for _, addr := range addrs {
@@ -98,85 +108,106 @@ func multiListen(
 		ml.listeners = append(ml.listeners, l)
 	}
 
-	for _, l := range ml.listeners {
+	for i, l := range ml.listeners {
+		i := i
 		ml.wg.Add(1)
+		ch := make(chan connErrPair, 1)
+		ml.connChs = append(ml.connChs, ch)
 		go func(l net.Listener) {
+			fmt.Printf("sub[%d]: starting\n", i)
 			defer ml.wg.Done()
 			for {
+				fmt.Printf("sub[%d]: accepting\n", i)
 				conn, err := l.Accept()
-				ml.mu.Lock()
-				if ml.closed {
-					ml.mu.Unlock()
+				pair := connErrPair{conn: conn, err: err}
+				fmt.Printf("sub[%d]: queueing %s\n", i, pair.String())
+
+				// Signal ml.Accept()
+				ch <- pair
+				select {
+				case ml.readyCh <- struct{}{}:
+					fmt.Printf("sub[%d]: sent ready\n", i)
+				case <-ml.stopCh:
+					fmt.Printf("sub[%d]: exiting\n", i)
+					close(ch)
 					return
 				}
-				// enqueue the accepted connection
-				ml.connErrQueue = append(ml.connErrQueue, connErrPair{conn: conn, err: err})
-
-				// signal the waiting ml.Accept() to consume accepted connection from the queue.
-				select {
-				case ml.acceptReadyCh <- struct{}{}:
-				default:
-				}
-				ml.mu.Unlock()
 			}
 		}(l)
 	}
 	return ml, nil
 }
 
-// Accept implements net.Listener.
-// It waits for and returns a connection from any of the sub-listener.
+// Accept implements net.Listener.  It waits for and returns a connection from
+// any of the sub-listener.
 func (ml *multiListener) Accept() (net.Conn, error) {
 	for {
-		// atomically return and remove the first element of the queue if it's not empty
-		ml.mu.Lock()
-		if len(ml.connErrQueue) > 0 {
-			connErr := ml.connErrQueue[0]
-			ml.connErrQueue = ml.connErrQueue[1:]
-			ml.mu.Unlock()
-			return connErr.conn, connErr.err
-		}
-		ml.mu.Unlock()
-
 		// wait for any sub-listener to enqueue an accepted connection
-		_, ok := <-ml.acceptReadyCh
-		if !ok {
-			// The "acceptReadyCh" channel will be closed only when Close() is called on the multiListener.
+		if _, ok := <-ml.readyCh; !ok {
+			// The "readyCh" channel will be closed only when Close() is called on the multiListener.
 			// Closing of this channel implies that all sub-listeners are also closed, which causes a
 			// "use of closed network connection" error on their Accept() calls. We return the same error
 			// for multiListener.Accept() if multiListener.Close() has already been called.
 			return nil, fmt.Errorf("use of closed network connection")
 		}
+
+		for _, ch := range ml.connChs {
+			select {
+			case connErr, ok := <-ch:
+				if !ok {
+					return nil, fmt.Errorf("use of closed network connection")
+				}
+				return connErr.conn, connErr.err
+			default: // nothing there
+			}
+		}
+		// If we get here we got a "ready" but no connection?
 	}
 }
 
-// Close implements net.Listener.
-// It will close all sub-listeners and wait for the go-routines to exit.
+// Close implements net.Listener.  It will close all sub-listeners and wait for
+// the go-routines to exit.
 func (ml *multiListener) Close() error {
-	ml.mu.Lock()
-	if ml.closed {
-		ml.mu.Unlock()
-		return fmt.Errorf("use of closed network connection")
-	}
-	ml.closed = true
-	close(ml.acceptReadyCh)
-	ml.mu.Unlock()
+	// Tell all sub-listeners to stop.
+	close(ml.stopCh)
 
 	// Closing the listeners causes Accept() to immediately return an error,
-	// which serves as the exit condition for the sub-listener go-routines.
+	// which triggers the exit condition for the sub-listener go-routines.
 	for _, l := range ml.listeners {
 		_ = l.Close()
 	}
 
+	// Drain any already-queued connections.
+	for i, ch := range ml.connChs {
+		fmt.Printf("draining [%d]\n", i)
+		for connErr := range ch {
+			fmt.Printf("draining [%d] => %v\n", i, connErr.String())
+			if connErr.conn != nil {
+				connErr.conn.Close()
+			}
+		}
+		fmt.Printf("drained [%d]\n", i)
+	}
+
 	// Wait for all the sub-listener go-routines to exit.
 	ml.wg.Wait()
+	close(ml.readyCh)
+
 	return nil
 }
 
-// Addr is an implementation of the net.Listener interface.
-// It always returns the address of the first listener.
-// Callers should  use conn.LocalAddr() to obtain the actual
-// local address of the sub-listener.
+// Addr is an implementation of the net.Listener interface.  It always returns
+// the address of the first listener.  Callers should  use conn.LocalAddr() to
+// obtain the actual local address of the sub-listener.
 func (ml *multiListener) Addr() net.Addr {
 	return ml.listeners[0].Addr()
 }
+
+// Addrs is like Addr, but returns the address for all registered listeners.
+func (ml *multiListener) Addrs() []net.Addr {
+	ret := []net.Addr{}
+	for _, l := range ml.listeners {
+		ret = append(ret, l.Addr())
+	}
+	return ret
+}
```

Reply:
If we used `reflect.Select` we could get rid of `readyCh` entirely and just select on the array of `connCh`

Reply:
yeah, let's not panic/recover inside a library if we can avoid it.  It is costly, if I recall, and I don't think we need it.  We just need clarity on our semantics.

As a general rule, Go asssumes that writers "own" channels and are responsible for closing them "when there's no more data".  It's safe to read a closed channel but not write.

If the semantic we want is approximately "each sub-listener has a bounded number of accepted connections waiting for the app to accept", then that suggests to me that we need a per-listener queue (channels are good at this).  If we prefer "the multi-listener has a bounded number of accepted connections waiting, regardless of which sub-listener", it can get even simpler, at the cost of potentially starving, which I guess can happen anyway unless we iterate the conn channels more "fairly" (e.g. keep track of which one we read last, and then resume there like a ring).  I don't think we need that.

Simpler form.  It may still race, but I think I got them all.  The fun one was between app calling Accept() in a tight loop and sub-listeners.  This presumes we don't yet change to the `AddListener()` model.


```diff
diff --git a/net/multi_listen.go b/net/multi_listen.go
index d508805..6672978 100644
--- a/net/multi_listen.go
+++ b/net/multi_listen.go
@@ -29,20 +29,26 @@ type connErrPair struct {
 	err  error
 }
 
+func (p connErrPair) String() string {
+	if p.conn != nil {
+		return p.conn.RemoteAddr().String()
+	}
+	if p.err != nil {
+		return p.err.Error()
+	}
+	return "neither conn nor err were valid"
+}
+
 // multiListener implements net.Listener
 type multiListener struct {
 	listeners []net.Listener
 
-	wg     sync.WaitGroup
-	mu     sync.Mutex
-	closed bool
+	wg sync.WaitGroup
 
-	// connErrQueue holds the connections accepted by sub-listeners
-	connErrQueue []connErrPair
-
-	// acceptReadyCh is used as a semaphore to wake up the waiting
-	// multiListener.Accept() when new connections are available
-	acceptReadyCh chan any
+	// connCh passes accepted connections, from child listeners to parent.
+	connCh chan connErrPair
+	// stopCh communicates from parent to child listeners.
+	stopCh chan struct{}
 }
 
 // compile time check to ensure *multiListener implements net.Listener
@@ -58,13 +64,13 @@ var _ net.Listener = &multiListener{}
 //  2. Use "tcp4" or "tcp6" to exclusively listen on IPv4 or IPv6 family, respectively.
 //  3. The host can accept names (e.g, localhost) and it will create a listener for at
 //     most one of the host's IP.
-func MultiListen(ctx context.Context, network string, addrs []string) (net.Listener, error) {
+func MultiListen(ctx context.Context, network string, addrs ...string) (net.Listener, error) {
+	var lc net.ListenConfig
 	return multiListen(
 		ctx,
 		network,
 		addrs,
 		func(ctx context.Context, network, address string) (net.Listener, error) {
-			var lc net.ListenConfig
 			return lc.Listen(ctx, network, address)
 		})
 }
@@ -78,14 +84,15 @@ func multiListen(
 	listenFunc func(ctx context.Context, network, address string) (net.Listener, error),
 ) (net.Listener, error) {
 	if !(network == "tcp" || network == "tcp4" || network == "tcp6") {
-		return nil, fmt.Errorf("network '%s' not supported", network)
+		return nil, fmt.Errorf("network %q not supported", network)
 	}
 	if len(addrs) == 0 {
 		return nil, fmt.Errorf("no address provided to listen on")
 	}
 
 	ml := &multiListener{
-		acceptReadyCh: make(chan any),
+		connCh: make(chan connErrPair),
+		stopCh: make(chan struct{}),
 	}
 
 	for _, addr := range addrs {
@@ -98,85 +105,94 @@ func multiListen(
 		ml.listeners = append(ml.listeners, l)
 	}
 
-	for _, l := range ml.listeners {
+	for i, l := range ml.listeners {
+		i := i
 		ml.wg.Add(1)
 		go func(l net.Listener) {
+			fmt.Printf("sub[%d]: starting\n", i) //DBG
 			defer ml.wg.Done()
 			for {
+				fmt.Printf("sub[%d]: accepting\n", i) //DBG
+				// Accept() is blocking, unless ml.Close() is called, in which
+				// case it will return immediately with an error.
 				conn, err := l.Accept()
-				ml.mu.Lock()
-				if ml.closed {
-					ml.mu.Unlock()
+				pair := connErrPair{conn: conn, err: err}
+				terminate := (err != nil)
+
+				select {
+				case ml.connCh <- pair:
+					fmt.Printf("sub[%d]: sent conn %s\n", i, pair.String()) //DBG
+				case <-ml.stopCh:
+					terminate = true
+				}
+				if terminate {
+					fmt.Printf("sub[%d]: exiting\n", i) //DBG
 					return
 				}
-				// enqueue the accepted connection
-				ml.connErrQueue = append(ml.connErrQueue, connErrPair{conn: conn, err: err})
-
-				// signal the waiting ml.Accept() to consume accepted connection from the queue.
-				select {
-				case ml.acceptReadyCh <- struct{}{}:
-				default:
-				}
-				ml.mu.Unlock()
 			}
 		}(l)
 	}
 	return ml, nil
 }
 
-// Accept implements net.Listener.
-// It waits for and returns a connection from any of the sub-listener.
+// Accept implements net.Listener.  It waits for and returns a connection from
+// any of the sub-listener.
 func (ml *multiListener) Accept() (net.Conn, error) {
-	for {
-		// atomically return and remove the first element of the queue if it's not empty
-		ml.mu.Lock()
-		if len(ml.connErrQueue) > 0 {
-			connErr := ml.connErrQueue[0]
-			ml.connErrQueue = ml.connErrQueue[1:]
-			ml.mu.Unlock()
-			return connErr.conn, connErr.err
-		}
-		ml.mu.Unlock()
-
-		// wait for any sub-listener to enqueue an accepted connection
-		_, ok := <-ml.acceptReadyCh
-		if !ok {
-			// The "acceptReadyCh" channel will be closed only when Close() is called on the multiListener.
-			// Closing of this channel implies that all sub-listeners are also closed, which causes a
-			// "use of closed network connection" error on their Accept() calls. We return the same error
-			// for multiListener.Accept() if multiListener.Close() has already been called.
-			return nil, fmt.Errorf("use of closed network connection")
-		}
+	// wait for any sub-listener to enqueue an accepted connection
+	connErr, ok := <-ml.connCh
+	if !ok {
+		// The channel will be closed only when Close() is called on the
+		// multiListener.  Closing of this channel implies that all
+		// sub-listeners are also closed, which causes a "use of closed
+		// network connection" error on their Accept() calls. We return the
+		// same error for multiListener.Accept() if multiListener.Close()
+		// has already been called.
+		return nil, fmt.Errorf("use of closed network connection")
 	}
+	return connErr.conn, connErr.err
 }
 
-// Close implements net.Listener.
-// It will close all sub-listeners and wait for the go-routines to exit.
+// Close implements net.Listener.  It will close all sub-listeners and wait for
+// the go-routines to exit.
 func (ml *multiListener) Close() error {
-	ml.mu.Lock()
-	if ml.closed {
-		ml.mu.Unlock()
-		return fmt.Errorf("use of closed network connection")
-	}
-	ml.closed = true
-	close(ml.acceptReadyCh)
-	ml.mu.Unlock()
+	// Tell all sub-listeners to stop.
+	close(ml.stopCh)
 
-	// Closing the listeners causes Accept() to immediately return an error,
-	// which serves as the exit condition for the sub-listener go-routines.
+	// Closing the listeners causes Accept() to immediately return an error in
+	// the sub-listener go-routines.
 	for _, l := range ml.listeners {
 		_ = l.Close()
 	}
 
 	// Wait for all the sub-listener go-routines to exit.
 	ml.wg.Wait()
+	close(ml.connCh)
+
+	// Drain any already-queued connections.
+	fmt.Printf("draining\n") //DBG
+	for connErr := range ml.connCh {
+		fmt.Printf("draining => %v\n", connErr.String()) //DBG
+		if connErr.conn != nil {
+			connErr.conn.Close()
+		}
+	}
+	fmt.Printf("drained\n") //DBG
+
 	return nil
 }
 
-// Addr is an implementation of the net.Listener interface.
-// It always returns the address of the first listener.
-// Callers should  use conn.LocalAddr() to obtain the actual
-// local address of the sub-listener.
+// Addr is an implementation of the net.Listener interface.  It always returns
+// the address of the first listener.  Callers should  use conn.LocalAddr() to
+// obtain the actual local address of the sub-listener.
 func (ml *multiListener) Addr() net.Addr {
 	return ml.listeners[0].Addr()
 }
+
+// Addrs is like Addr, but returns the address for all registered listeners.
+func (ml *multiListener) Addrs() []net.Addr {
+	ret := []net.Addr{}
+	for _, l := range ml.listeners {
+		ret = append(ret, l.Addr())
+	}
+	return ret
+}
```

Reply:
I tried this approach panic/race is still there. 

Added a unit test to mock race conditions.
```shell
go test k8s.io/utils/net -run TestMultiListen_Accept/accept_some_connections -v --count 10
```
Have tried `reflect.Select` with one channel per sub-listener, race is still there (but the test rarely panics now), will try to come up with a reliable solution.

Reply:
Running your test against https://github.com/kubernetes/utils/pull/311#discussion_r1655283651 I get no fails, even with 1M runs.  I tried to tickle the races by forcing a sleep in strategic places, but I can't make it happen.  I added some comments.

It only closes `connCh` AFTER the waitgroup is done, so sub-listeners can't write to it after close, and nobody ever writes to `stopCh`, it's just used as a `select` case, so sub-listeners don't block on sending a connection.

https://github.com/kubernetes/utils/compare/master...thockin:kubernetes-utils:pr-311-plus-thockin?expand=1